hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java

```diff
@@ -156,21 +157,55 @@ public boolean recoverLease(Path f) throws IOException {
       throw e;
     }
 
-    // finalize the final block and get block length
-    List<OmKeyLocationInfo> locationInfoList = leaseKeyInfo.getKeyInfo().getLatestVersionLocations().getLocationList();
-    if (!locationInfoList.isEmpty()) {
-      OmKeyLocationInfo block = locationInfoList.get(locationInfoList.size() - 1);
+    OmKeyLocationInfoGroup keyLatestVersionLocations = leaseKeyInfo.getKeyInfo().getLatestVersionLocations();
+    List<OmKeyLocationInfo> keyLocationInfoList = keyLatestVersionLocations.getLocationList();
+    OmKeyLocationInfoGroup openKeyLatestVersionLocations = leaseKeyInfo.getOpenKeyInfo().getLatestVersionLocations();
+    List<OmKeyLocationInfo> openKeyLocationInfoList = openKeyLatestVersionLocations.getLocationList();
+
+    int openKeyLocationSize = openKeyLocationInfoList.size();
+    int keyLocationSize = keyLocationInfoList.size();
+    OmKeyLocationInfo openKeyFinalBlock = null;
+    OmKeyLocationInfo openKeyPenultimateBlock = null;
+    OmKeyLocationInfo keyFinalBlock;
+
+    if (keyLocationSize > 0) {
+      // Block info from fileTable
+      keyFinalBlock = keyLocationInfoList.get(keyLocationSize - 1);
+      // Block info from openFileTable
+      if (openKeyLocationSize > 1) {
+        openKeyFinalBlock = openKeyLocationInfoList.get(openKeyLocationSize - 1);
+        openKeyPenultimateBlock = openKeyLocationInfoList.get(openKeyLocationSize - 2);
+      } else if (openKeyLocationSize > 0) {
+        openKeyFinalBlock = openKeyLocationInfoList.get(0);
+      }
+      // Finalize the final block and get block length
       try {
-        block.setLength(getAdapter().finalizeBlock(block));
+        // CASE 1: When openFileTable has more block than fileTable
```

Comment:
does the test case cover all three cases?

Reply:
the rest looks good.

Reply:
Thanks for the review, Case1 and 2 is covered in the new and existing test case. One more test case I have added now to cover Case3.